File: problems/cpp/validate-binary-search-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

bool is_valid_bst(TreeNode* root, long min_key, long max_key)
{
    if (!root)
        return true;

    return root->val > min_key && root->val < max_key
           && is_valid_bst(root->left, min_key, root->val)
           && is_valid_bst(root->right, root->val, max_key);
}

bool isValidBST(TreeNode* root)
{
    // Need to use long here because of a few testcases with -2**31 and (2**31 - 1)
    // valued keys
    return is_valid_bst(root, numeric_limits<long>::min(), numeric_limits<long>::max());
}

File: code/python/validate-binary-search-tree.py
# 83 ms, 02/06/2018
# Time  : O(n)
# Space : O(h)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def are_keys_within_bounds(root, lb, ub):
            if not root:
                return True
            return (lb < root.val < ub
                    and are_keys_within_bounds(root.left, lb, root.val)
                    and are_keys_within_bounds(root.right, root.val, ub))
        
        return are_keys_within_bounds(root, float('-inf'), float('+inf'))
